In [1]:
# --- Imports ---
import pandas as pd

# --- Load main merged DataFrame ---
ords_prods_merge = pd.read_pickle('/Users/cem/Desktop/Data Immersion/Achievement 4_Project/Data/Prepared Data/ords_prods_merged_labeled.pkl')

# create a subset
df = ords_prods_merge[:1000000]

df.shape

(1000000, 19)

In [2]:
df.head(10)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   
5   2398795        1    prior             2          3                  7   
6   2398795        1    prior             2          3                  7   
7   2398795        1    prior             2          3                  7   
8   2398795        1    prior             2          3                  7   
9   2398795        1    prior             2          3                  7   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                     0.0       14084                  2          0   
2                     0.0       12427                  3          0   
3                     0.0       26088                  4          0   
4                     0.0       26405                  5          0   
5                    15.0         196                  1          1   
6                    15.0       10258                  2          0   
7                    15.0       12427                  3          1   
8                    15.0       13176                  4          0   
9                    15.0       26088                  5          1   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda      77.0            7.0     9.0   
1  Organic Unsweetened Vanilla Almond Milk      91.0           16.0    12.5   
2                      Original Beef Jerky      23.0           19.0     4.4   
3               Aged White Cheddar Popcorn      23.0           19.0     4.7   
4         XL Pick-A-Size Paper Towel Rolls      54.0           17.0     1.0   
5                                     Soda      77.0            7.0     9.0   
6                               Pistachios     117.0           19.0     3.0   
7                      Original Beef Jerky      23.0           19.0     4.4   
8                   Bag of Organic Bananas      24.0            4.0    10.3   
9               Aged White Cheddar Popcorn      23.0           19.0     4.7   

  _merge        price_label     busiest_day      day_type  \
0   both  Mid-range product  Least busy day  Average days   
1   both  Mid-range product  Least busy day  Average days   
2   both  Low-range product  Least busy day  Average days   
3   both  Low-range product  Least busy day  Average days   
4   both  Low-range product  Least busy day  Average days   
5   both  Mid-range product  Least busy day  Slowest days   
6   both  Low-range product  Least busy day  Slowest days   
7   both  Low-range product  Least busy day  Slowest days   
8   both  Mid-range product  Least busy day  Slowest days   
9   both  Low-range product  Least busy day  Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders  
5        Average orders  
6        Average orders  
7        Average orders  
8        Average orders  
9        Average orders

In [3]:
df.groupby('product_name')

In [4]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1.0              14.794722
2.0              17.091743
3.0              17.929713
4.0              17.891973
5.0              15.214270
6.0              15.382228
7.0              17.699986
8.0              16.485269
9.0              15.965921
10.0             20.091818
11.0             16.484474
12.0             15.615845
13.0             16.485279
14.0             17.508672
15.0             15.690354
16.0             18.005083
17.0             16.155822
18.0             19.606536
19.0             17.630236
20.0             17.138204
21.0             21.996844

In [6]:
df.groupby('department_id').order_number.mean()

department_id
1.0     14.794722
2.0     17.091743
3.0     17.929713
4.0     17.891973
5.0     15.214270
6.0     15.382228
7.0     17.699986
8.0     16.485269
9.0     15.965921
10.0    20.091818
11.0    16.484474
12.0    15.615845
13.0    16.485279
14.0    17.508672
15.0    15.690354
16.0    18.005083
17.0    16.155822
18.0    19.606536
19.0    17.630236
20.0    17.138204
21.0    21.996844
Name: order_number, dtype: float64

In [7]:
## Q2

dept_order_means = ords_prods_merge.groupby('department_id')['order_number'].mean()
print(dept_order_means)


department_id
1.0     15.457687
2.0     17.277920
3.0     17.179756
4.0     17.811403
5.0     15.215751
6.0     16.439806
7.0     17.225773
8.0     15.340520
9.0     15.895474
10.0    20.197148
11.0    16.170371
12.0    15.887622
13.0    16.583304
14.0    16.759763
15.0    16.165037
16.0    17.663250
17.0    15.694469
18.0    19.310397
19.0    17.177343
20.0    16.473447
21.0    22.902379
Name: order_number, dtype: float64


In [8]:
## Q3
# Mean order_number per department (subset)
subset_means = df.groupby('department_id')['order_number'].mean()

# Mean order_number per department (full dataset)
full_means = ords_prods_merge.groupby('department_id')['order_number'].mean()

comparison = pd.DataFrame({
    'Subset (1M rows)': subset_means,
    'Full Dataset': full_means
})
print(comparison)

               Subset (1M rows)  Full Dataset
department_id                                
1.0                   14.794722     15.457687
2.0                   17.091743     17.277920
3.0                   17.929713     17.179756
4.0                   17.891973     17.811403
5.0                   15.214270     15.215751
6.0                   15.382228     16.439806
7.0                   17.699986     17.225773
8.0                   16.485269     15.340520
9.0                   15.965921     15.895474
10.0                  20.091818     20.197148
11.0                  16.484474     16.170371
12.0                  15.615845     15.887622
13.0                  16.485279     16.583304
14.0                  17.508672     16.759763
15.0                  15.690354     16.165037
16.0                  18.005083     17.663250
17.0                  16.155822     15.694469
18.0                  19.606536     19.310397
19.0                  17.630236     17.177343
20.0                  17.138204   

## Analysis: Mean Order Number by Department (Subset vs. Full Dataset)

- The **subset (1 million rows)** provides a rough estimate of the mean order_number for each department, but it's based on a partial view of the data.
- The **full dataset** includes all available rows and gives a more accurate, complete view.
- For some departments, the differences between subset and full dataset are small, which suggests that the subset is reasonably representative.
- This highlights the importance of using the full dataset when performing final analyses, especially for departments that aren't well represented in the subset.
  


In [10]:
## Q4
ords_prods_merge['avg_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform('mean')

ords_prods_merge.loc[ords_prods_merge['avg_order'] > 20, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['avg_order'] <= 20) & (ords_prods_merge['avg_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['avg_order'] <= 10, 'loyalty_flag'] = 'New customer'

ords_prods_merge['loyalty_flag'].value_counts()

loyalty_flag
New customer        12090686
Loyal customer      10944939
Regular customer     9399434
Name: count, dtype: int64

In [11]:
## Q5

ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10944278.0  10.466255  340.951098  1.0  4.2  7.4  11.2   
New customer      12089896.0  13.312824  592.351781  1.0  4.2  7.4  11.3   
Regular customer   9398856.0  12.032798  510.208181  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    99999.0  
New customer      99999.0  
Regular customer  99999.0

In [12]:
ords_prods_merge.groupby('loyalty_flag')['prices'].mean()

loyalty_flag
Loyal customer      10.466255
New customer        13.312824
Regular customer    12.032798
Name: prices, dtype: float64

In [13]:
ords_prods_merge.groupby('loyalty_flag')['prices'].median()

loyalty_flag
Loyal customer      7.4
New customer        7.4
Regular customer    7.4
Name: prices, dtype: float64

## Price Analysis by Customer Loyalty

I analyzed whether loyal customers buy more expensive products compared to new or regular customers.

### Key Observations:
- **Median product price** is exactly the same (7.4) across all customer groups, indicating that the typical product purchased is consistent.
- **New customers** have the highest mean price (~13.31), suggesting they may experiment with more expensive items during early usage.
- **Regular customers** follow with a mean of aroud 12.03, while loyal customers have the lowest mean (~10.47), possibly reflecting a tendency to repurchase lower-priced or familiar products.
- This pattern may indicate that loyalty leads to more consistent and cost-conscious purchasing, whereas new users explore more premium options.
- The marketing team could use this insight to offer higher-priced product recommendations or promotions to new users, while incentivizing bulk deals or discounts for loyal customers.


In [17]:
## Q6
ords_prods_merge['avg_price_per_user'] = ords_prods_merge.groupby('user_id')['prices'].transform('mean')
ords_prods_merge.loc[ords_prods_merge['avg_price_per_user'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['avg_price_per_user'] >= 10, 'spending_flag'] = 'High spender'
ords_prods_merge['spending_flag'].value_counts()



spending_flag
Low spender     31799519
High spender      635540
Name: count, dtype: int64

In [18]:
## Q7
ords_prods_merge['median_days'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform('median')

ords_prods_merge.loc[ords_prods_merge['median_days'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[
    (ords_prods_merge['median_days'] > 10) & (ords_prods_merge['median_days'] <= 20),
    'order_frequency_flag'
] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

ords_prods_merge['order_frequency_flag'].value_counts()


order_frequency_flag
Frequent customer        22816041
Regular customer          6929012
Non-frequent customer     2690006
Name: count, dtype: int64

## Creating Order Frequency Flags

I analyzed customer behavior based on the median number of days between their orders. Users were labeled as:

- **Frequent customer**: median ≤ 10 days
- **Regular customer**: 10 < median ≤ 20 days
- **Non-frequent customer**: median > 20 days

This helps the Instacart team determine how regularly each customer shops and can be used to send personalized notifications for reordering.
